### PR TITLE
Add README for DataLoader2

### PR DESCRIPTION
**Pull request by ejguan**
Add brief description of `DataLoader2` and related components.

### CODE REVIEW

File: torchdata/dataloader2/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+- dynamic sharding for multi-process and distributed data loading
+- multiple backend `ReadingServices` (WIP for `DistributedReadingService`)
+- `DataPipe` graph in-place modification like shuffle control, memory pinning, etc.
+- snapshot the state of data-preprocessing pipeline (WIP)
+
+## Why DataLoader2
+
+`DataPipe` is introduced to not only decompose the data pre-processing operations, but also decouple the overloaded
+data-manipulation features from `DataLoader`. So, a light-weight `DataLoader2` is required. Besides, a certain features
+can only be achieved with `DataLoader2` like snapshotting and bridging data-processing pipeline between research and
```
Comment (NivekT) [inline torchdata/dataloader2/README.md:14]:
```suggestion
data-manipulation features from `DataLoader`. So, a light-weight `DataLoader2` is required. Besides, certain features
can only be achieved with `DataLoader2`, such as snapshotting and bridging data-processing pipeline between research and
```

Comment (NivekT) [reply torchdata/dataloader2/README.md:14]: I am actually not sure what feature you are referring to by "bridging data-processing pipeline between research and". What feature is that one?

File: torchdata/dataloader2/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+production.
+
+These options are configured by the constructor arguments of a `DataLoader2`, which has signature:
```
Comment (NivekT) [inline torchdata/dataloader2/README.md:17]:
```suggestion
These options are configured by the constructor arguments of `DataLoader2`, which has the signature:
```

File: torchdata/dataloader2/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+DataLoader2(datapipe, datapipe_adapter_fn=None, reading_service=None)
+```
+
+The sections below describe in details the effects and usages of these options.
+
+## DataPipe
+
+`DataLoader2` supports both `MapDataPipe` and `IterDataPipe`. To understand the basic structure of `DataPipe`, please
+see [What are DataPipes?](https://github.com/pytorch/data#what-are-datapipes)
+
```
Comment (NivekT) [inline torchdata/dataloader2/README.md:29]: It may be worth mentioning that it doesn't support `Dataset` and `IterableDataset`? Unless they put a wrapper around it?

File: torchdata/dataloader2/README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+## Data-Processing Options (datapipe_adapter_fn)
```
Comment (NivekT) [inline torchdata/dataloader2/README.md:30]:
```suggestion
## Data-Processing Options (`datapipe_adapter_fn`)
```

File: torchdata/dataloader2/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+`Adapters` are used to configure, modify and extend the graph of `DataPipes` in `DataLoader2`. For example,
+`Shuffle(False)` can be provided to `DataLoader2`, which would disable any of `shuffle` operations in the graph of
+`DataPipes`.
+
+And, we will provide more `Adapters` to cover data-processing options such as `PinMemory`. If you have feature requests
+about the `Adapters` you'd like to be provided, please open a GitHub issue.
+
+For specific needs, `DataLoader2` accepts any custom `Adapter` as long as it inherits from the `Adapter` class.
+
```
Comment (NivekT) [inline torchdata/dataloader2/README.md:40]: Do we have more potentially `Adapter` aside from `Shuffle` and `PinMemory`? I think it will help users to understand the concept better if we give more potential use cases.

File: torchdata/dataloader2/README.md (hunk 1 above)
Comment (NivekT) [reply torchdata/dataloader2/README.md:14]: I think one feature you have in mind is `ReadingService`? If so we can directly mention it here.

**Review by NivekT — APPROVED** [review]
Mostly LGTM! Just some nit comments for clarity.

Adding a quick code snippet with `DataLoader2` in use with `MultiprocessingReadService` may be helpful for users as well, but we can also wait until it further matures and add a complete tutorial then.

File: torchdata/dataloader2/README.md (hunk 1 above)
Comment (ejguan) [reply torchdata/dataloader2/README.md:14]: Let me rephrase it.

File: torchdata/dataloader2/README.md (hunk 3 above)
Comment (ejguan) [reply torchdata/dataloader2/README.md:29]: Sounds good.
